qiskit/ml/algorithms/quantum_kernel.py

```diff
@@ -0,0 +1,272 @@
+            # |<0|Psi^daggar(y) x Psi(x)|0>|^2, take the amplitude
+            tmp = np.vdot(v_a, v_b)
+            kernel_value = np.vdot(tmp, tmp).real  # pylint: disable=no-member
+        else:
+            result = results.get_counts(idx)
+
+            measurement_basis = '0' * self._feature_map.num_qubits
+            kernel_value = result.get(measurement_basis, 0) / sum(result.values())
+        return kernel_value
+
+    def evaluate(self, x_vec, y_vec=None):
+        """
+        Construct kernel matrix for given data and feature map
+
+        If y is None, self inner product is calculated.
+        If using `statevector_simulator`, only build circuits for  Phi(x)|0>, then perform inner
+        product classically.
+
+        Args:
+            x_vec (numpy.ndarray): 2D array of datapoints, NxD, where N is the number of datapoints,
+                               D is the feature dimension
+            y_vec (numpy.ndarray): 2D array of datapoints, MxD, where M is the number of datapoints,
+                               D is the feature dimension
+
+        Returns:
+            numpy.ndarray: 2-D matrix, NxM
+
+        Raises:
+            ValueError:
+                - x_vec and/or y_vec are not two dimensional arrays
+                - x_vec and/or y_vec have incompatible dimension with feature map
+        """
+        if x_vec.ndim != 2:
+            raise ValueError("x_vec must be a 2D array")
+
+        if y_vec is not None and y_vec.ndim != 2:
+            raise ValueError("y_vec must be a 2D array")
+
+        if x_vec.shape[1] != self._feature_map.num_qubits:
+            raise ValueError("x_vec and class feature map incompatible dimensions.\n" +
+                             "x_vec has %s dimensions, but feature map has %s." %
+                             (x_vec.shape[1], self._feature_map.num_qubits))
+
+        if y_vec is not None and y_vec.shape[1] != self._feature_map.num_qubits:
+            raise ValueError("y_vec and class feature map incompatible dimensions.\n" +
+                             "y_vec has %s dimensions, but feature map has %s." %
+                             (y_vec.shape[1], self._feature_map.num_qubits))
+
+        # determine if calculating self inner product
+        is_symmetric = False
+        if y_vec is None:
+            y_vec = x_vec
+        if np.array_equal(x_vec, y_vec):
+            is_symmetric = True
+
+        # initialise kernel matrix
+        kernel = np.zeros((x_vec.shape[0], y_vec.shape[0]))
+
+        # set diagonal to 1 if symmetric
+        if is_symmetric:
+            np.fill_diagonal(kernel, 1)
+
+        # get indices to calculate
+        if is_symmetric:
+            mus, nus = np.triu_indices(x_vec.shape[0], k=1)  # remove diagonal
+        else:
+            mus, nus = np.indices((x_vec.shape[0], y_vec.shape[0]))
+            mus = np.asarray(mus.flat)
+            nus = np.asarray(nus.flat)
+
+        is_statevector_sim = self._quantum_instance.is_statevector
```

Comment:
I think the evaluate method should check that a QuantumInstance has been set and raise an appropriate error. By default the constructor sets this to None and this line will end up failing. (This level of checking was done in QuantumAlgo run() method in the past, which this explicitly did not subclass).

Reply:
Added in 6192f8d